Fix ConcurrentModificationException in parser error handler

henrymai opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

Exception:
```
Unexpected failure when handling parsing error. This is likely a bug in the implementation
java.util.ConcurrentModificationException
        at java.base/java.util.HashMap$HashIterator.nextNode(HashMap.java:1597)
        at java.base/java.util.HashMap$KeyIterator.next(HashMap.java:1620)
        at io.trino.sql.parser.ErrorHandler$Analyzer.process(ErrorHandler.java:216)
```

This was happening because after the first iteration of the while loop,
`endTokens` is a reference to `nextTokens`, so when we iterate through
`endTokens`, we are actually iterating through `nextTokens`. So since we are
iterating through `nextTokens` and also adding to `nextTokens` at the same
time, we end up with a `ConcurrentModificationException`.


<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues



<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

(X) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

henrymai commented:
Sorry for all the updates. Just minor simplifications.

henrymai commented on core/trino-parser/src/main/java/io/trino/sql/parser/ErrorHandler.java:
```diff
@@ -211,14 +211,11 @@ public Result process(ATNState currentState, int tokenIndex, RuleContext context
-                    nextTokens.addAll(process(new ParsingState(nextState, endToken, false, parser), 0));
-                }
+                final ATNState nextState = ((RuleTransition) atn.states.get(context.invokingState).transition(0)).followState;
+                endTokens = endTokens.stream().flatMap(endToken ->
```
Currently the trino-parser test fails but I'm not sure if the test itself is wrong.

Meaning, I can change line 216 to mimic the old behavior to give us the expected result for the test:
```
                endTokens = Stream.concat(endTokens.stream(), endTokens.stream().flatMap(endToken ->
                    process(new ParsingState(nextState, endToken, false, parser), 0).stream())).collect(Collectors.toSet());
```

The test in question that fails is:
`io.trino.sql.parser.TestSqlParserErrorHandling.testStatement(String, String)[46]`

Where the expected output is:
```
<"line 1:21: mismatched input 'SELECT'. Expecting: '%', '(', ')', '*', '+', ',', '-', '.', '/', 'AND', 'AS', 'AT', 'EXCEPT', 'FETCH', 'FROM', 'GROUP', 'HAVING', 'INTERSECT', 'LIMIT', 'OFFSET', 'OR', 'ORDER', 'SELECT', 'TABLE', 'UNION', 'VALUES', 'WHERE', 'WINDOW', '[', '||', <EOF>, <identifier>, <predicate>">
```

And where my current output is:
```
<"line 1:21: mismatched input 'SELECT'. Expecting: '%', ')', '*', '+', ',', '-', '.', '/', 'AND', 'AS', 'AT', 'EXCEPT', 'FETCH', 'FROM', 'GROUP', 'HAVING', 'INTERSECT', 'LIMIT', 'OFFSET', 'OR', 'ORDER', 'UNION', 'WHERE', 'WINDOW', '[', '||', <identifier>, <predicate>">
```

The diff looks like this:
![test_diff](https://user-images.githubusercontent.com/136857/217172932-3f9840af-321d-406d-a043-f97eaf20fbee.png)

pettyjamesm reviewed (approved):
LGTM

cc: @martint

martint commented on core/trino-parser/src/main/java/io/trino/sql/parser/ErrorHandler.java:
```diff
@@ -211,14 +211,11 @@ public Result process(ATNState currentState, int tokenIndex, RuleContext context
-                    ATNState nextState = ((RuleTransition) atn.states.get(context.invokingState).transition(0)).followState;
-                    nextTokens.addAll(process(new ParsingState(nextState, endToken, false, parser), 0));
-                }
+                final ATNState nextState = ((RuleTransition) atn.states.get(context.invokingState).transition(0)).followState;
```
Don't use `final` for local variable.

henrymai replied:
Done

martint commented on core/trino-parser/src/main/java/io/trino/sql/parser/ErrorHandler.java:
```diff
@@ -211,14 +211,11 @@ public Result process(ATNState currentState, int tokenIndex, RuleContext context
-                    nextTokens.addAll(process(new ParsingState(nextState, endToken, false, parser), 0));
-                }
+                final ATNState nextState = ((RuleTransition) atn.states.get(context.invokingState).transition(0)).followState;
+                endTokens = endTokens.stream().flatMap(endToken ->
```
Format as:

```
endTokens = endTokens.stream()
        .flatMap(endToken -> process(new ParsingState(nextState, endToken, false, parser), 0).stream())
        .collect(Collectors.toSet());
```

henrymai replied:
Done

martint reviewed (commented):
Thanks! Just a couple of minor comments.

martint commented:
Please squash the commits and change the commit message to "Fix ConcurrentModificationException in parser error handler".

Also, can you add a test that reproduces the error?

henrymai commented:
> Please squash the commits and change the commit message to "Fix ConcurrentModificationException in parser error handler".
> 

Thanks for the review, I've addressed those comments above.

> Also, can you add a test that reproduces the error?

Unfortunately this is a very rare problem that we've run into in production and we have not been able to come up with a synthetic reproduction of this problem. I know this isn't a great answer, but at least it is fairly clear that this is an actual bug from just looking at the code.
